### PR TITLE
Add setting for only showing unread rooms

### PR DESCRIPTION
**Pull request by andybalaam**
Under Appearance -> Advanced, add a setting "Only show rooms with unread messages" which filters out rooms that have no unread messages in them.

rec-2022-01-20_14 18 12

Fixes https://github.com/vector-im/element-web/issues/7185
Fixes https://github.com/vector-im/element-web/issues/16009

This is a proposal rather than a final design.  I'd like feedback from @matrix-org/design on this.

I actually think a more-visible setting, perhaps like https://vk.com/messenger has, would probably be better, since switching back and forth between the two views may be a useful workflow.

---
Here's what your changelog entry will look like:

## ✨ Features
 * Add setting for only showing unread rooms ([\#7589](https://github.com/matrix-org/matrix-react-sdk/pull/7589)). Fixes vector-im/element-web#7185 and vector-im/element-web#16009. Contributed by @andybalaam.

### CODE REVIEW

File: src/components/views/settings/tabs/user/AppearanceUserSettingsTab.tsx [hunk 1]
```diff
@@ -119,6 +119,7 @@ export default class AppearanceUserSettingsTab extends React.Component<IProps, I
             );
             advanced = <>
                 <SettingsFlag name="useCompactLayout" level={SettingLevel.DEVICE} useCheckbox={true} />
+                <SettingsFlag name="showOnlyUnreadRooms" level={SettingLevel.DEVICE} useCheckbox={true} />
```
Comment (turt2live) [inline src/components/views/settings/tabs/user/AppearanceUserSettingsTab.tsx:122]: This should probably go with the other room list settings. Possibly even be a per-sublist setting to match it up with the ordering constraints.

**Review by turt2live — APPROVED** [review]
Overall the code looks good to me, though I do suspect design will want to take a look at this in depth.

File: src/components/views/settings/tabs/user/AppearanceUserSettingsTab.tsx (hunk 1 above)
Comment (andybalaam) [reply src/components/views/settings/tabs/user/AppearanceUserSettingsTab.tsx:122]: I really hate the per-sublist settings.  I'd love for Design to get their teeth into the roomlist!

File: src/stores/room-list/filters/UnreadFilterCondition.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
+/*
+Copyright 2020, 2021 The Matrix.org Foundation C.I.C.
```
Comment (t3chguy) [inline src/stores/room-list/filters/UnreadFilterCondition.ts:2]: wrong year

File: src/stores/room-list/filters/UnreadFilterCondition.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+import { RoomNotificationStateStore } from "../../notifications/RoomNotificationStateStore";
+
+/**
+ * A filter condition for the room list which reveals rooms of a particular
+ * name, or associated name (like a room alias).
+ */
+export class UnreadFilterCondition extends EventEmitter implements IFilterCondition {
+    constructor() {
+        super();
+    }
```
Comment (t3chguy) [inline src/stores/room-list/filters/UnreadFilterCondition.ts:30]: redundant

**Comment by t3chguy** [comment]
How would a user navigate to a room via search with this thing enabled if it isn't unread?

**Review by t3chguy — CHANGES_REQUESTED** [review]
See above

**Comment by andybalaam** [comment]
> How would a user navigate to a room via search with this thing enabled if it isn't unread?

With this change, the room search still shows up things that match your search even if they aren't unread.

But, thanks to your comment, I noticed that when searching, all rooms with unread message show up as if they matched the search, which is wrong!  So I need to fix that.

File: src/stores/room-list/filters/UnreadFilterCondition.ts (hunk 2 above)
Comment (andybalaam) [reply src/stores/room-list/filters/UnreadFilterCondition.ts]: Fixed in d75d516bc9

File: src/stores/room-list/filters/UnreadFilterCondition.ts (hunk 3 above)
Comment (andybalaam) [reply src/stores/room-list/filters/UnreadFilterCondition.ts]: Fixed in d75d516bc9

**Comment by t3chguy** [comment]
Interesting, I did not expect the filters to UNION, the docs don't specify

**Comment by madduck** [comment]
I this expected to work? I cannot find the setting on `develop.element.io`…

**Comment by ajkessel** [comment]
I would really like a button above the room list to toggle between "unread" and "all", which is, I think, the gist of this issue, and would provide functionality that other IM clients have (including Beeper). It doesn't seem like it would be a heavy implementation lift. I see this issue is closed, though, so has the idea been definitively rejected?

**Pull request closed without merging**